router/core/transport.go

```diff
@@ -78,6 +89,107 @@ func (ct *CustomTransport) RoundTrip(req *http.Request) (*http.Response, error)
 	return resp, err
 }
 
+type responseWithBody struct {
+	res  *http.Response
+	body []byte
+}
+
+func (ct *CustomTransport) allowSingleFlight(req *http.Request) bool {
+	if ct.sf == nil {
+		// Single flight is disabled
+		return false
+	}
+
+	if req.Header.Get("Upgrade") != "" {
+		// Websocket requests are not idempotent
+		return false
+	}
+
+	if req.Header.Get("Accept") == "text/event-stream" {
+		// SSE requests are not idempotent
+		return false
+	}
+
+	if resolve.SingleFlightDisallowed(req.Context()) {
+		// Single flight is disallowed for this request (e.g. because it is a Mutation)
+		return false
+	}
+
+	return true
+}
+
+func (ct *CustomTransport) roundTripSingleFlight(req *http.Request) (*http.Response, error) {
+
+	if !ct.allowSingleFlight(req) {
+		return ct.roundTripper.RoundTrip(req)
+	}
+
+	keyGen := pool.Hash64.Get()
+	defer pool.Hash64.Put(keyGen)
+
+	if req.Body != nil {
+		body, err := io.ReadAll(req.Body)
+		if err != nil {
+			return nil, err
+		}
+		_, _ = keyGen.Write(body)
+		req.Body = io.NopCloser(bytes.NewReader(body))
+	}
+
+	unsortedHeaders := make([]string, 0, len(req.Header))
+
+	for key := range req.Header {
+		value := req.Header.Get(key)
+		unsortedHeaders = append(unsortedHeaders, key+value)
+	}
+
+	sort.Strings(unsortedHeaders)
+	for i := range unsortedHeaders {
+		_, _ = keyGen.Write(unsafebytes.StringToBytes(unsortedHeaders[i]))
+	}
+
+	sum := keyGen.Sum64()
+	key := strconv.FormatUint(sum, 10)
+
+	v, err, shared := ct.sf.Do(key, func() (interface{}, error) {
+		res, err := ct.roundTripper.RoundTrip(req)
+		if err != nil {
+			return nil, err
+		}
+		body, err := io.ReadAll(res.Body)
+		if err != nil {
+			return nil, err
+		}
+		return &responseWithBody{
+			res:  res,
+			body: body,
+		}, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	sfStats := resolve.GetSingleFlightStats(req.Context())
+	if sfStats != nil {
+		sfStats.SingleFlightUsed = true
+		sfStats.SingleFlightSharedResponse = shared
+	}
+	rwb := v.(*responseWithBody)
+	res := &http.Response{}
```

Comment:
Some props are missing like proto*, close, TLS Check the struct definition on core.